torchtext/datasets/text_classification.py

```diff
@@ -113,36 +113,88 @@ def get_vocab(self):
+        tokenizer = get_tokenizer("basic_english")
 
+    if isinstance(data_select, str):
+        data_select = [data_select]
+    if not set(data_select).issubset(set(('train', 'test'))):
+        raise TypeError('Given data selection {} is not supported!'.format(data_select))
+
+    dataset_tar = download_from_url(URLS[dataset_name], root=root)
+    extracted_files = extract_archive(dataset_tar)
+
+    iters_group = {}
+    if 'train' in data_select:
+        iters_group['vocab'] = _imdb_iterator('train', extracted_files, tokenizer, ngrams)
+    for item in data_select:
+        iters_group[item] = _imdb_iterator(item, extracted_files,
+                                           tokenizer, ngrams, yield_cls=True)
+    return iters_group
+
+
+def _setup_datasets(iters_group, vocab, removed_tokens, data_select):
+
+    if isinstance(data_select, str):
+        data_select = [data_select]
     if vocab is None:
-        logging.info('Building Vocab based on {}'.format(train_csv_path))
-        vocab = build_vocab_from_iterator(_csv_iterator(train_csv_path, ngrams))
+        if 'vocab' not in iters_group.keys():
+            raise TypeError("Must pass a vocab if train is not selected.")
+        logging.info('Building Vocab based on train data')
+        vocab = build_vocab_from_iterator(iters_group['vocab'])
     else:
         if not isinstance(vocab, Vocab):
             raise TypeError("Passed vocabulary is not of type Vocab")
     logging.info('Vocab has {} entries'.format(len(vocab)))
-    logging.info('Creating training data')
-    train_data, train_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(train_csv_path, ngrams, yield_cls=True), include_unk)
-    logging.info('Creating testing data')
-    test_data, test_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(test_csv_path, ngrams, yield_cls=True), include_unk)
-    if len(train_labels ^ test_labels) > 0:
-        raise ValueError("Training and test labels don't match")
-    return (TextClassificationDataset(vocab, train_data, train_labels),
-            TextClassificationDataset(vocab, test_data, test_labels))
-
-
-def AG_NEWS(*args, **kwargs):
+
+    data = {}
+    for item in iters_group.keys():
+        data[item] = {}
+        data[item]['data'] = []
+        data[item]['labels'] = []
+        logging.info('Creating {} data'.format(item))
+        data_iter = _create_data_from_iterator(vocab, iters_group[item], removed_tokens)
+        for cls, tokens in data_iter:
+            data[item]['data'].append((torch.tensor(cls),
+                                       torch.tensor([token_id for token_id in tokens])))
+            data[item]['labels'].append(cls)
+        data[item]['labels'] = set(data[item]['labels'])
+
+    return tuple(TextClassificationDataset(vocab, data[item]['data'],
+                                           data[item]['labels']) for item in data_select)
+
+
+def AG_NEWS(root='.data', ngrams=2, vocab=None, removed_tokens=[],
```

Comment:
To avoid typos as might come along with wide-ranging changes or simply maintaining something as this long list of arguments and default values I suggest to create a new function that abstracts the function call. 

```
    return _setup_datasets(generate_data_iterators('AG_NEWS', root, ngrams,
                                                   tokenizer, data_select),
                           vocab, removed_tokens, data_select)
```

Reply:
Add a function `_initiate_datasets` to handle the two branches.